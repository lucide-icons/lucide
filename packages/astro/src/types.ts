import type { HTMLAttributes } from 'astro/types';

// Type that the Astro language server needs to infer component props in Astro files
export type AstroComponent = (_props: IconProps) => any;

export interface IconProps extends SVGAttributes {
  color?: string;
  size?: number | string;
  'stroke-width'?: number | string;
  absoluteStrokeWidth?: boolean;
  class?: string;
  iconNode?: IconNode;
}

export type SVGAttributes = HTMLAttributes<'svg'>;

export type IconNode = IconNodeChild[];

type IconNodeChild = [elementName: SVGElements, attrs: SVGAttributes];

// All possible svg elements according to the Astro definitions
type SVGElements =
  | 'svg'
  | 'animate'
  | 'circle'
  | 'clipPath'
  | 'defs'
  | 'desc'
  | 'ellipse'
  | 'feBlend'
  | 'feColorMatrix'
  | 'feComponentTransfer'
  | 'feComposite'
  | 'feConvolveMatrix'
  | 'feDiffuseLighting'
  | 'feDisplacementMap'
  | 'feDistantLight'
  | 'feFlood'
  | 'feFuncA'
  | 'feFuncB'
  | 'feFuncG'
  | 'feFuncR'
  | 'feGaussianBlur'
  | 'feImage'
  | 'feMerge'
  | 'feMergeNode'
  | 'feMorphology'
  | 'feOffset'
  | 'fePointLight'
  | 'feSpecularLighting'
  | 'feSpotLight'
  | 'feTile'
  | 'feTurbulence'
  | 'filter'
  | 'foreignObject'
  | 'g'
  | 'image'
  | 'line'
  | 'linearGradient'
  | 'marker'
  | 'mask'
  | 'metadata'
  | 'path'
  | 'pattern'
  | 'polygon'
  | 'polyline'
  | 'radialGradient'
  | 'rect'
  | 'stop'
  | 'switch'
  | 'symbol'
  | 'text'
  | 'textPath'
  | 'tspan'
  | 'use'
  | 'view';
