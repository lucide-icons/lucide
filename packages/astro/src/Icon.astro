---
import defaultAttributes from './defaultAttributes';
import type { IconProps as Props } from './types';

const {
  color = 'currentColor',
  size = 24,
  'stroke-width': strokeWidth = 2,
  absoluteStrokeWidth = false,
  iconNode = [],
  class: className,
  ...rest
} = Astro.props;
---

<svg
  {...{
    ...defaultAttributes,
    width: size,
    height: size,
    stroke: color,
    'stroke-width': absoluteStrokeWidth ? (Number(strokeWidth) * 24) / Number(size) : strokeWidth,
    ...rest,
  }}
  class:list={['lucide', className]}
>
  {iconNode.map(([Tag, attrs]) => <Tag {...attrs} />)}
  <slot />
</svg>
