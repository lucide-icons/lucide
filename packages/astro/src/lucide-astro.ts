export * from './icons/index';
export * as icons from './icons/index';
export * from './aliases';
export * from './types';

export { default as defaultAttributes } from './defaultAttributes';
export { default as createLucideIcon } from './createLucideIcon';
export { default as Icon } from './Icon.astro';
