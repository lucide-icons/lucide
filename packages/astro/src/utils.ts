/**
 * Converts string to kebab case
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string): string =>
  string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

/**
 * Merges classes into a single string
 *
 * @param {array} classes
 * @returns {string} A string of classes
 */
export const mergeClasses = <ClassType = string | undefined | null>(
  ...classes: ClassType[]
): string =>
  classes
    .filter((className, index, array) => {
      return (
        Boolean(className) &&
        (className as string).trim() !== '' &&
        array.indexOf(className) === index
      );
    })
    .join(' ')
    .trim();
