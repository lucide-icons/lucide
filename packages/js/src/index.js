import icons from './icons';
import toSvg from './to-svg';
import replace from './replace';

module.exports = { icons, toSvg, replace };
