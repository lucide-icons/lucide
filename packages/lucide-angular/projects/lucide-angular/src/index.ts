
export * from './lib/lucide-angular.component'
export * from './lib/lucide-angular.module';
export * from './helpers/create-element'

export * from './icons'
export * as icons from './icons';
