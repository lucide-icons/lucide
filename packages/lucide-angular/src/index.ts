import * as icons from './icons';

export * from './lib/lucide-angular.component';
export * from './lib/lucide-angular.module';
export * from './helpers/create-element';
export * from './icons';
export { icons };
