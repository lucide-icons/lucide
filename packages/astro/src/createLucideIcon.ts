import { mergeClasses, toKebabCase } from './utils';
import type { AstroComponentFactory } from 'astro/runtime/server/render/astro/factory.js';
import type { IconNode } from './types';
import { render, renderSlot, createComponent, renderComponent } from 'astro/compiler-runtime';
import Icon from './Icon.astro';

export default (iconName: string, iconNode: IconNode): AstroComponentFactory => {
  const Component = createComponent(
    ($$result, $$props: Record<string, any>, $$slots) => {
      const { class: className, ...restProps } = $$props;
      return render`${renderComponent(
        $$result,
        'Icon',
        Icon,
        {
          class: mergeClasses(
            Boolean(iconName) && `lucide-${toKebabCase(iconName)}`,
            Boolean(className) && className,
          ),
          iconNode,
          ...restProps,
        },
        { default: () => render`${renderSlot($$result, $$slots['default'])}` },
      )}`;
    },
    undefined,
    'none',
  );
  return Component;
};
