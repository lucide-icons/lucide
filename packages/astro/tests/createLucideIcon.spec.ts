import { describe, it, expect } from 'vitest';
import { createLucideIcon } from '../src/lucide-astro';
import { airVent } from './testIconNodes';
import { render } from './utils';

describe('Using createLucideIcon', () => {
  it('should create a component from an iconNode', async () => {
    const AirVent = createLucideIcon('AirVent', airVent);
    const { container } = await render(AirVent);

    expect(container.innerHTML).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });
});
