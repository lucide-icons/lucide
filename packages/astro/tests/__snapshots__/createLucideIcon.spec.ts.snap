// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Using createLucideIcon > should create a component from an iconNode 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     width="24"
     height="24"
     viewbox="0 0 24 24"
     fill="none"
     stroke="currentColor"
     stroke-width="2"
     stroke-linecap="round"
     stroke-linejoin="round"
     class="lucide lucide-air-vent"
>
  <path d="M6 12H4a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v5a2 2 0 0 1-2 2h-2">
  </path>
  <path d="M6 8h12">
  </path>
  <path d="M18.3 17.7a2.5 2.5 0 0 1-3.16 3.83 2.53 2.53 0 0 1-1.14-2V12">
  </path>
  <path d="M6.6 15.6A2 2 0 1 0 10 17v-5">
  </path>
</svg>
`;
